Source/Core/Core/IOS/USB/USB_KBD.cpp

```diff
@@ -315,11 +168,143 @@ u8 USB_KBD::m_KeyCodesAZERTY[256] = {
     0x00,  // ' '
     0x38,  // '!'
     0x00,  // Nothing interesting past this point.
-
 };
 #else
-u8 USB_KBD::m_KeyCodesQWERTY[256] = {0};
+constexpr std::array<u8, 256> s_key_codes_qwerty{};
+
+constexpr std::array<u8, 256> s_key_codes_azerty{};
+#endif
+}  // Anonymous namespace
+
+USB_KBD::MessageData::MessageData(MessageType type, u8 modifiers, PressedKeyData pressed_keys)
+    : msg_type{Common::swap32(static_cast<u32>(type))}, modifiers{modifiers}, pressed_keys{
+                                                                                  pressed_keys}
+{
+}
+
+// TODO: support in netplay/movies.
+
+USB_KBD::USB_KBD(Kernel& ios, const std::string& device_name) : Device(ios, device_name)
+{
+}
+
+IPCCommandResult USB_KBD::Open(const OpenRequest& request)
+{
+  INFO_LOG(IOS, "USB_KBD: Open");
+  IniFile ini;
+  ini.Load(File::GetUserPath(F_DOLPHINCONFIG_IDX));
+  ini.GetOrCreateSection("USB Keyboard")->Get("Layout", &m_keyboard_layout, KBD_LAYOUT_QWERTY);
+
+  m_message_queue = {};
+  m_old_key_buffer.fill(false);
+  m_old_modifiers = 0x00;
+
+  // m_message_queue.emplace(MessageType::KeyboardConnect, 0, PressedKeyData{});
```

Comment:
This one had been commented out before, but do we know why? The code itself seems to make sense here...

Reply:
Looking back through the commit history, it seems it was commented out almost a decade ago at:

https://github.com/dolphin-emu/dolphin/commit/98ce4c4d8f9ae1cc68e1f18281ad0ea010e115db#diff-30816fa5ef7811ae9b07868599c46042R42

The commit message doesn't explain why though. Perhaps it was commented out for testing, but never re-enabled. I don't know enough about the workings of the keyboard driver to make that judgment call though.

Reply:
Yep, it's been disabled forever. It's a shame none of this is documented properly.